### PR TITLE
Fix for #2377 : DLNA/UPnP - Basedir fix in DLNA broadcasts

### PR DESCRIPTION
**Pull request by BaronGreenback**
There seems to be two issues regarding DLNA, and one with Uri parsing (affecting dlna and web streaming being a proxy)

- First is that it ignored the value in BaseDir, meaning that if i was reverse proxying in - the DLNA broadcast message would include a potentially unconnectable URL. BaseDir is now included as part of the message.

eg. http://192.168.1.207:8096/dlna/fc0f0986a2c94b60bd0d60563d3fe61a/connectionmanager/connectionmanager.xml

with a basedir of "web" becomes

http://192.168.1.207:8096/web/dlna/fc0f0986a2c94b60bd0d60563d3fe61a/connectionmanager/connectionmanager.xml

- Second, it didn't support alternative hostnames. It's now been modified to use the command line argument "published-server-url" as an override in the publication string.

so starting jellyfin with "--published-server-url http://jellyfin.org.uk" will cause http://jellyfin.org.uk/dlna/.... to be advertised.

and starting jellyfin with "--published-server-url http://jellyfin.org.uk" with a basedir of "web" will cause http://jellyfin.org.uk/web/dlna/.... to be advertised.

Possible future addition (would require #3257) would be to advertise different uri's over the internal/external interfaces.

Also fixes a proxy bug whereby UriHelper is not aware of the X-Forwarded-Proto header.

### CODE REVIEW

File: Emby.Dlna/Main/DlnaEntryPoint.cs [hunk 1]
```diff
@@ -99,6 +103,8 @@ public class DlnaEntryPoint : IServerEntryPoint, IRunBeforeStartup
             _networkManager = networkManager;
             _logger = loggerFactory.CreateLogger<DlnaEntryPoint>();
 
+            _publishedServerUrl = conf["PublishedServerUrl"];
```
Comment (cvium) [inline Emby.Dlna/Main/DlnaEntryPoint.cs:106]: We really should be using https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-3.1#bind-hierarchical-configuration
```suggestion
            _publishedServerUrl = configuration["PublishedServerUrl"];
```

File: Emby.Dlna/Main/DlnaEntryPoint.cs [hunk 2]
```diff
@@ -80,7 +83,8 @@ public class DlnaEntryPoint : IServerEntryPoint, IRunBeforeStartup
             ISocketFactory socketFactory,
             INetworkManager networkManager,
             IUserViewManager userViewManager,
-            ITVSeriesManager tvSeriesManager)
+            ITVSeriesManager tvSeriesManager,
+            IConfiguration conf)
```
Comment (cvium) [inline Emby.Dlna/Main/DlnaEntryPoint.cs:87]:
```suggestion
            IConfiguration configuration)
```

File: Emby.Dlna/Main/DlnaEntryPoint.cs [hunk 3]
```diff
@@ -276,8 +282,9 @@ private async Task RegisterServerEndpoints()
 
                 _logger.LogInformation("Registering publisher for {0} on {1}", fullService, address);
 
-                var descriptorUri = "/dlna/" + udn + "/description.xml";
-                var uri = new Uri(_appHost.GetLocalApiUrl(address) + descriptorUri);
+                var descriptorUri = (_config.Configuration.BaseUrl ?? string.Empty + "/dlna/" + udn + "/description.xml").TrimStart('/');
```
Comment (cvium) [inline Emby.Dlna/Main/DlnaEntryPoint.cs:285]: This is a bit messy imo

File: Emby.Dlna/Main/DlnaEntryPoint.cs [hunk 4]
```diff
@@ -276,8 +282,9 @@ private async Task RegisterServerEndpoints()
...
+
+                var uri = new Uri(string.IsNullOrEmpty(_publishedServerUrl) ? _appHost.GetLocalApiUrl(address) + "/" + descriptorUri : _publishedServerUrl + descriptorUri);
```
Comment (cvium) [inline Emby.Dlna/Main/DlnaEntryPoint.cs:287]: I am not a fan of ternaries in string concatenation.

File: Emby.Dlna/Main/DlnaEntryPoint.cs (hunk 1 above)
Comment (BaronGreenback) [reply Emby.Dlna/Main/DlnaEntryPoint.cs:106]: agree

File: Emby.Dlna/Main/DlnaEntryPoint.cs (hunk 4 above)
Comment (BaronGreenback) [reply Emby.Dlna/Main/DlnaEntryPoint.cs:288]: extracted out

File: Emby.Dlna/Main/DlnaEntryPoint.cs (hunk 3 above)
Comment (BaronGreenback) [reply Emby.Dlna/Main/DlnaEntryPoint.cs:285]: cleaned it up.

**Comment by PrplHaz4** [comment]
Tested this quickly tonight and I'm still seeing the IP being advertised in `SSDP NOTIFY`.  I've set the environment variable `JELLYFIN_PublishedServerUrl` to `'https://jftest.mydomain.com'` and verified it exists inside the container.  Do any of the other network settings need to be changed for this to take?

```
Internet Protocol Version 4, Src: jftest.mydomain.com (192.168.1.141), Dst: 239.255.255.250 (239.255.255.250)
User Datagram Protocol, Src Port: 44787, Dst Port: 1900
Simple Service Discovery Protocol
    NOTIFY * HTTP/1.1\r\n
    HOST: 239.255.255.250:1900\r\n
    DATE: Mon, 03 Aug 2020 00:17:37 GMT\r\n
    CACHE-CONTROL: max-age = 1800\r\n
    LOCATION: http://192.168.1.141:8096/dlna/af9f06335f4d490685987391b5f6b4be/description.xml\r\n
    SERVER: Linux/Unix 4.14.138.0 UPnP/1.0 RSSDP/1.0\r\n
    NTS: ssdp:alive\r\n
    NT: uuid:af9f06335f4d490685987391b5f6b4be\r\n
    USN: uuid:af9f06335f4d490685987391b5f6b4be\r\n
    \r\n
    [Full request URI: http://239.255.255.250:1900*]
```

**Comment by BaronGreenback** [comment]
I think has to be a command line arg and not an environment var. (Start JF with --PublishedServerUrl). Don't know why - that's just the way it is.

**Comment by PrplHaz4** [comment]
> 
> 
> I think it has to be a command line arg and not an environment var.
> 
> Start JF with --PublishedServerUrl

I had to change to a skewered argument name (`--published-server-url`) and adjust my docker-compose entrypoint:
```
entrypoint: ["./jellyfin/jellyfin", "--datadir", "/config", "--cachedir", "/cache", "--ffmpeg", "/usr/lib/jellyfin-ffmpeg/ffmpeg", "--published-server-url", "https://jftest.mydomain.com"]
```

Now I can see the correct URL in the NOTIFY messages, but for some reason, VLC is not showing jf as a DLNA server (but a 10.6.2 instance on the same network does)...maybe a problem for another day/build...

Great work on this one - I think all of your networking changes are going to make a lot of people happy...

```
Simple Service Discovery Protocol
    NOTIFY * HTTP/1.1\r\n
    HOST: 239.255.255.250:1900\r\n
    DATE: Mon, 03 Aug 2020 15:33:21 GMT\r\n
    CACHE-CONTROL: max-age = 1800\r\n
    LOCATION: https://jftest.mydomain.com/dlna/af9f06335f4d490685987391b5f6b4be/description.xml\r\n
    SERVER: Linux/Unix 4.14.138.0 UPnP/1.0 RSSDP/1.0\r\n
    NTS: ssdp:alive\r\n
    NT: upnp:rootdevice\r\n
    USN: uuid:af9f06335f4d490685987391b5f6b4be::upnp:rootdevice\r\n
    \r\n
    [Full request URI: http://239.255.255.250:1900*]
```

**Comment by BaronGreenback** [comment]
A soupedup version of this is in the new network backend. 

It's going to put out as an Alpha for testing before the acceptance of the PR - as there are just too many changes.

**Comment by BaronGreenback** [comment]
As for detection - I've had similar. 

Switch on blast alive messages and reduce the time down. 

I find if the app is run after the initial alive message message is sent, it's another 30 mins before it will be detected.

**Comment by PrplHaz4** [comment]
Trying to figure out this discovery issue - my train of thought was around the DLNA clients not being able to resolve domain names or following redirects...when I set `published-server-url` back to `http://192.168.1.141:8096`, DLNA discovery works quickly, just as it did pre-10.6.  This will bypass the RP and not use SSL...

Here is a potential bug in the DIDL returned - it looks like the wrong protocol is being specified - if `published-server-url` is set to `https://jftest2.mydomain.com`, then URLBase returned in the XML is `http://jftest2.mydomain.com` (http instead of https).

<details>
<summary>Show https://jftest2.mydomain.com/dlna/af9f06335f4d490685987391b5f6b4be/description.xml</summary>

```xml
<root>
	<specVersion>
		<major>
			1
		</major>
		<minor>
			0
		</minor>
	</specVersion>
	<URLBase>
		http://jftest2.mydomain.com
	</URLBase>
	<device>
		<X_DLNACAP />
		<X_DLNADOC>
			DMS-1.50
		</X_DLNADOC>
		<X_DLNADOC>
			M-DMS-1.50
		</X_DLNADOC>
		<deviceType>
			urn:schemas-upnp-org:device:MediaServer:1
		</deviceType>
		<friendlyName>
			Jellyfin - jftest2
		</friendlyName>
		<manufacturer>
			Jellyfin
		</manufacturer>
		<manufacturerURL>
			https://github.com/jellyfin/jellyfin
		</manufacturerURL>
		<modelDescription>
			UPnP/AV 1.0 Compliant Media Server
		</modelDescription>
		<modelName>
			Jellyfin Server
		</modelName>
		<modelNumber>
			01
		</modelNumber>
		<modelURL>
			https://github.com/jellyfin/jellyfin
		</modelURL>
		<serialNumber>
			af9f06335f4d490685987391b5f6b4be
		</serialNumber>
		<UPC />
		<UDN>
			uuid:af9f06335f4d490685987391b5f6b4be
		</UDN>
		<iconList>
			<icon>
				<mimetype>
					image/png
				</mimetype>
				<width>
					240
				</width>
				<height>
					240
				</height>
				<depth>
					24
				</depth>
				<url>
					/dlna/af9f06335f4d490685987391b5f6b4be/icons/logo240.png
				</url>
			</icon>
			<icon>
				<mimetype>
					image/jpeg
				</mimetype>
				<width>
					240
				</width>
				<height>
					240
				</height>
				<depth>
					24
				</depth>
				<url>
					/dlna/af9f06335f4d490685987391b5f6b4be/icons/logo240.jpg
				</url>
			</icon>
			<icon>
				<mimetype>
					image/png
				</mimetype>
				<width>
					120
				</width>
				<height>
					120
				</height>
				<depth>
					24
				</depth>
				<url>
					/dlna/af9f06335f4d490685987391b5f6b4be/icons/logo120.png
				</url>
			</icon>
			<icon>
				<mimetype>
					image/jpeg
				</mimetype>
				<width>
					120
				</width>
				<height>
					120
				</height>
				<depth>
					24
				</depth>
				<url>
					/dlna/af9f06335f4d490685987391b5f6b4be/icons/logo120.jpg
				</url>
			</icon>
			<icon>
				<mimetype>
					image/png
				</mimetype>
				<width>
					48
				</width>
				<height>
					48
				</height>
				<depth>
					24
				</depth>
				<url>
					/dlna/af9f06335f4d490685987391b5f6b4be/icons/logo48.png
				</url>
			</icon>
			<icon>
				<mimetype>
					image/jpeg
				</mimetype>
				<width>
					48
				</width>
				<height>
					48
				</height>
				<depth>
					24
				</depth>
				<url>
					/dlna/af9f06335f4d490685987391b5f6b4be/icons/logo48.jpg
				</url>
			</icon>
		</iconList>
		<presentationURL>
			http://jftest2.mydomain.com/web/index.html
		</presentationURL>
		<serviceList>
			<service>
				<serviceType>
					urn:schemas-upnp-org:service:ContentDirectory:1
				</serviceType>
				<serviceId>
					urn:upnp-org:serviceId:ContentDirectory
				</serviceId>
				<SCPDURL>
					/dlna/af9f06335f4d490685987391b5f6b4be/contentdirectory/contentdirectory.xml
				</SCPDURL>
				<controlURL>
					/dlna/af9f06335f4d490685987391b5f6b4be/contentdirectory/control
				</controlURL>
				<eventSubURL>
					/dlna/af9f06335f4d490685987391b5f6b4be/contentdirectory/events
				</eventSubURL>
			</service>
			<service>
				<serviceType>
					urn:schemas-upnp-org:service:ConnectionManager:1
				</serviceType>
				<serviceId>
					urn:upnp-org:serviceId:ConnectionManager
				</serviceId>
				<SCPDURL>
					/dlna/af9f06335f4d490685987391b5f6b4be/connectionmanager/connectionmanager.xml
				</SCPDURL>
				<controlURL>
					/dlna/af9f06335f4d490685987391b5f6b4be/connectionmanager/control
				</controlURL>
				<eventSubURL>
					/dlna/af9f06335f4d490685987391b5f6b4be/connectionmanager/events
				</eventSubURL>
			</service>
		</serviceList>
	</device>
</root>
```
</details>

**Comment by BaronGreenback** [comment]
Good catch - will go and find out why

**Comment by BaronGreenback** [comment]
This section of the code parses the url it receives for the info. It's done in at least 18 places - so before i go further - i'm assuming the following is happened?

Your setup is something like:-

(internet) -> proxy -> jellyfin

The pubished-server-url is used from the web to get to the proxy, whereby it's forwarded over http to jellyfin - possibly by ip address, with the url stored in a header?

If the above is correct, then the following is happening:-

Jellyfin is parsing the url correctly, but as the url is calculated from the incoming url (http://jftest2.mydomain.com/web/index.html) it's returning the info as a http and not https.

The fix - would be to do some more checking/correcting in DescriptionXmlBuilder with the serverAddress property.

**Comment by PrplHaz4** [comment]
> 
> 
> This section of the code parses the url it receives for the info. It's done in at least 18 places - so before i go further - i'm assuming the following is happened?
> 
> Your setup is something like:-
> 
> (internet) -> proxy -> jellyfin
> 
> The pubished-server-url is used from the web to get to the proxy, whereby it's forwarded over http to jellyfin - possibly by ip address, with the url stored in a header? This last doesn't really matter as it's working. It's the http part that is relevant.
> 
> If the above is correct, then the following is happening:-
> 
> Jellyfin is parsing the url correctly, but as the url is calculated from the incoming url (http://jftest2.mydomain.com/web/index.html) it's returning the info as a http and not https.
> 
> The fix - would be to do some more checking/correcting in DescriptionXmlBuilder with the serverAddress property.
> 
> If you can confirm - then i'll edit this PR with a fix.

Pretty close - for me it's all local network so far though:

> local net -> traefik proxy (https) -> jellyfin (http only)

The proxy forwards to an IP on the internal docker network (`http://172.17.0.1:8096`) - so I'm guessing it somehow translates that back to a hostname?  Or is it passing the hostname but setting the protocol explicitly to http because https is not enabled on JF (like we saw in #3712)...

Also note, I'm not a DLNA expert, so not sure what the exact rules are - the [spec doc](https://openconnectivity.org/upnp-specs/UPnP-arch-DeviceArchitecture-v2.0-20200417.pdf) seems to imply that domain names are ok in LOCATION:

> **`LOCATION`**
> Required.  Field  value  contains  a  URL  to  the  UPnP  description  of  the  root  device.  Normally  the  host  portion  contains  a  literal  IP  address  rather  than  a  domain name  in  unmanaged  networks.  Specified  by  UPnP  vendor.  Single absolute URL (see RFC 3986)

> **`<URLBase>`**
> The use of URLBase element is deprecated by this specification. UPnP 2.0  devices shall notinclude  URLBase  in  their  description  documents.  To  ensure  interoperability  with  UPnP  1.0  devices,  control  points  shall be  able  to  process  URLBase  if  it  is  specified  and  use  it  for  resolving relative URLs that appear elsewhere in the description. If relative URLs are included in the device description, control points shall resolve them into absolute URLs in accordance with clause 5 of RFC 3986, using either URLBase (if specified) or the location from which the device  description  was  retrieved  as  the  base  URL,  before  using  these  URLs  for  their  respective purposes.

> **`<presentationURL>`**
> Recommended.  URL  to  presentation  for  device  (see  clause  5,  “Presentation”). shall  be relative  to  the  URL  at  which  the  device  description  is  located  in  accordance  with  the  rules  specified in clause 5 of RFC 3986. Specified by UPnP vendor. Single UR

**Comment by BaronGreenback** [comment]
As i understand it the url is embedded into a header by the proxy, so that the receiving app has some idea what the original url is.

X-Forwarded-For - The client IP
X-Forwarded-Host - The Host header from the client (e.g. www.example.com:80)
X-Forwarded-Proto - The protocol (e.g. HTTPS)

Traced the call back to WebSocketSharpRequest.cs method AbsoluteUri.

This calls Request.GetDisplayUrl() which doesn't seem to honour the X-Forwarded-Proto - hence the issue.

From looking at this - it could be a bug across the board - with potentially any of jellyfin's functionality that consumes the url.

Am delving a bit deeper - then will update this PR.

**Comment by BaronGreenback** [comment]
Checked UriHelper.GetDisplayUrl source code, and it isn't proxy aware. Have made an additional change in this PR to rectify this.

Could you check please and let me know pls? (i'm deep within a bug hunt my end - so can test at the moment.)

**Comment by PrplHaz4** [comment]
> 
> 
> Checked UriHelper.GetDisplayUrl source code, and it isn't proxy aware. Have made an additional change in this PR to rectify this.
> 
> Could you check please and let me know pls? (i'm deep within a bug hunt my end - so can test at the moment.)
> 
> This bug will affects streaming as well as dlna.

haha...guess hacking around it can only get so far....

I'll try to test this some time today - should I merge this over #3257 or just build it directly?

**Pull request closed without merging**